### ISSUE TITLE
[PREMIUM] Functions Host causes Container Exit/Restart

### ISSUE THREAD

**Issue by marcd123** [Azure/azure-functions-host#6985]
We have a Premium Python Durable Functions app that occasionally "hangs" for a very long time in between scheduled tasks. 

We've investigated the issue quite a bit and discovered that the hanging is due to our container exiting.

We regularly test the resource usage of the Function App in KUDU by using the top command. While in KUDU we notice this message in the bottom tray when we experience the long hang:

![hang-message](https://user-images.githubusercontent.com/9528376/102260704-6e563900-3ed6-11eb-9039-14fd2d214a24.png)

So the container has evidently exited or restarted. When we run the "Container Crash" report in the "Diagnose and Solve problems" tab of our function app, we can see the following error message about the container exit:

```
Container exited unexpectedly: last 10 seconds logs [2020-12-15T17:29:01.710719291Z /azure-functions-host/start.sh: line 28: 18 Killed 
```

The most notable portion is "18 Killed"

```
azure-functions-host/start.sh: line 28: 18 Killed 
```

After more investigation it seems like "18" is referring to Process ID 18 which is the Microsoft+ process on our container, which I'm guessing is the functions host:

<img width="1185" alt="KUDU Process 18" src="https://user-images.githubusercontent.com/9528376/102261098-d278fd00-3ed6-11eb-9f1a-a6f196d88a27.png">

For some reason the functions host is killing this process, and then the container restarts immediately after - so we believe these are related.

Why does the functions host kill this process, and is this the cause for the container restarting? Is there anything we can do to fix this issue?

#### Investigative information

Please provide the following:

- Timestamp: 2020-12-15T17:29:01.710719291Z
- Function App version: ~3
- Functions Host Version: 3.0.15149.0
- Function App name: labelright-test-v2
- Function name(s): can happen during any function invocation
- Invocation ID: not related to any particular function invocation
- Region: East US

#### Repro steps

It seems to occur randomly, either under heavy or light load it can happen.

The only way to know its happening is to be watching KUDU during regular load, or run the "Container Crash" report in the function app panel to see when it last occurred.

#### Expected behavior

The functions host does not kill the Microsoft+ process and does not cause a container restart.

#### Actual behavior

The functions host kills the Microsoft+ process and causes a container restart.

#### Related information 

* Premium Function App Plan EP2 (2 vCPU, 7GB RAM)
* Only scaling out to one instance on this plan
* Using Python 3.6
* Using a custom docker image, inherits from the Azure python-3.6-appservice image
* Only additions to our container are installing Java and some libraries for working with PDFs
* This is a Durable Functions app that has a mix of I/O and CPU bound tasks
* The application converts PDFs to high quality images, uses tensorflow models for object detection, and uses OpenCV for computer vision and image analysis tasks
* FUNCTIONS_WORKER_PROCESS_COUNT = 8 (issue occurs with this setting at 2 as well)

Tagging @davidmrdavid from a previous conversation we had on the Durable Functions repo about the same topic:

https://github.com/Azure/azure-functions-durable-extension/issues/1573#issuecomment-730568055

**Comment by davidmrdavid** [comment]
@pragnagopa: I've been following this case (in other threads) for a while. Could you take a look at the customer's questions above ^ . Thanks! :)  I've seen a handful of customers experience this "gap" in function executions when using Durable Function for Python, and maybe it's related to this `18 Killed ` message that the customer's discovered.

**Comment by marcd123** [comment]
I enabled App Service logs for this function app, and now I can see this in the logs right around the time the container exits. Not sure if it is before or after the exit. Also not sure if this could be the cause or just a result of the exit.

```
Heartbeat took longer than "00:00:01" at "12/15/2020 22:09:17 +00:00". This could be caused by thread pool starvation.
```

**Comment by davidmrdavid** [comment]
Hi @marcd123,

I looked at some internal logs from your application yesterday and, from what I gathered, your container was killed due to it consuming too much memory. It would also appear that the message:

```
azure-functions-host/start.sh: line 28: 18 Killed 
```

Indeed refers to the Functions Host process in the container. 

**So I think your theory checks out!** 
This would also explain the "gaps" in work completion you were seeing, because we spend some time trying to clean-up and re-create your container when this kind of exception occurs.

As for next steps, I have some ideas.
1. It appears that you have `FUNCTIONS_WORKER_PROCESS_COUNT` set to `2`.
This might explain the memory consumption we're seeing. If you have an ML model loaded in each process (or deal with really large data),  both of which live in the same container, that's probably going to blow-up memory-wise.

So a first step would be to reduce that setting to `1` , such that you only have 1 process per container, giving you more room for memory-expensive operations.

2. If you need to increase concurrency (I assume that's why you set  `FUNCTIONS_WORKER_PROCESS_COUNT` to `2`), we recently released a new setting called `PYTHON_THREADPOOL_THREAD_COUNT` to control the number of threads a Python Functions worker can use to manage function invocations. It is `1` by default, and goes up to `32`. I'd say start small with values like `2` or `4` and see if this improves your performance. Do note that going too large with this value could negatively impact performance due to context-switching.

We also have an upcoming documentation page that describes when to tweak `FUNCTIONS_WORKER_PROCESS_COUNT`  versus `PYTHON_THREADPOOL_THREAD_COUNT` but, due to the holidays coming up, it's been delayed in our docs-publishing pipeline. As soon as it's out, I'll try to link it here so you can check it out ⚡ ⚡ 

**In summary:**
- Your theory so far appears correct: the container exit would explain the delays, and the `18 killed` message appears to correspond to terminating a resource-hungry container
- As next steps, reduce `FUNCTIONS_WORKER_PROCESS_COUNT` to `1`. Then, after testing that for a bit, consider increasing  `FUNCTIONS_WORKER_PROCESS_COUNT` in case reducing the first setting had a performance impact.

After some testing, please let me know if this addresses your issue! I might also be unavailable for a week or so due to the holidays, just a heads up ❄️ ! Please let me know how this goes though 😄

**Comment by marcd123** [comment]
Hey @davidmrdavid 

Thanks for the follow up.

This past week I did a lot of experimenting and found the best settings for our app. I think we could close the issue, though I would like to write up the experience here so others with similar issues can search it up.

There were two specific bugs I was trying to fix:

1. Language Worker Process is killed via SIGKILL (Exit code 137) due to out-of-memory
2. Container is restarted causing a long pause in execution

The first issue I resolved in two steps. First, I found which parts of our app consumed the most memory and tried to reduce memory consumption as best as possible. Then, I was still experiencing language workers being killed at 8 workers so I started reducing the count. At 2 language workers, we were unable to produce the first issue above. I found it really helpful to force our function app scale to a single VM, and use KUDU to SSH in and the "top" command to view the language workers and how much memory they were using. Being on a single VM also helped ensure all traffic was going to the same pool of resources, and simulated the "worse case scenario" of heavy work going to one VM, even if you can scale out to many.

In regard to the container restarting, I think out-of-memory could be part of the issue, but this seemed to happen even under very light load. I was able to reproduce the container restart issue using a low-memory workflow with 8, 6, and even 4 language workers. The issue wasn't totally resolved until we dropped language workers to 2. This is the same number of vCPU on our EP2 service plan. So while memory could be part of the issue, I think it's possible that the system experiences CPU process locks if you use more language workers than you have vCPUs on your service plan. If you're using Node or Python Azure Functions, I wouldn't recommend using more language workers than you have vCPUs in your service plan (if you're on premium) - even if your workload is low on resource usage.

I'm glad to hear about the PYTHON_THREADPOOL_THREAD_COUNT, and can't wait to see the docs on this. Does this setting help a single language worker execute multiple function invocations in parallel, or does it just give a single function invocation access to more threads to work on?

PS: Even though we've had to work out a lot of issues to get things running just right, the application is very performant now and we're loving Azure Functions. When we move to production, we're thinking of running our function app in Kubernetes to expand beyond the VM sizes offered in the Elastic Premium plan and scale to more VMs.

Great work, great product!

**Comment by davidmrdavid** [comment]
Closing this issue as it was resolved!

@marcd123, the new performance docs are available here: https://docs.microsoft.com/en-us/azure/azure-functions/python-scale-performance-reference

Reach out again if you need anything! ⚡ ⚡

**Comment by marcd123** [comment]
Thanks David!

Marc DeMory
Emerging Technology Consultant
Accenture Liquid Studio - Chicago<https://in.accenture.com/liquidstudios/chicagoliquidstudios/>
m: 630-244-9625
***@***.***
________________________________
From: David Justo ***@***.***>
Sent: Thursday, March 11, 2021 1:02 PM
To: Azure/azure-functions-host ***@***.***>
Cc: DeMory, Marc ***@***.***>; Mention ***@***.***>
Subject: [External] Re: [Azure/azure-functions-host] [PREMIUM] Functions Host causes Container Exit/Restart (#6985)

This message is from an EXTERNAL SENDER - be CAUTIOUS, particularly with links and attachments.
________________________________


Closing this issue as it was resolved!

@marcd123<https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_marcd123&d=DwMFaQ&c=eIGjsITfXP_y-DLLX0uEHXJvU8nOHrUK8IrwNKOtkVU&r=BbThQ-iApsLvCBycuzvhkO8FJNgUkO6tTOl2tJGWpHk&m=MqDA26oNjO8DUewvFmarqGTHJFy6DKFnZE15mMN7p2A&s=j047zT98LlRh4oUXyjCe-K6lMWso7kGnR33iF82nk-M&e=>, the new performance docs are available here: https://docs.microsoft.com/en-us/azure/azure-functions/python-scale-performance-reference<https://urldefense.proofpoint.com/v2/url?u=https-3A__docs.microsoft.com_en-2Dus_azure_azure-2Dfunctions_python-2Dscale-2Dperformance-2Dreference&d=DwMFaQ&c=eIGjsITfXP_y-DLLX0uEHXJvU8nOHrUK8IrwNKOtkVU&r=BbThQ-iApsLvCBycuzvhkO8FJNgUkO6tTOl2tJGWpHk&m=MqDA26oNjO8DUewvFmarqGTHJFy6DKFnZE15mMN7p2A&s=kBI4C4db8rN4vOZ-80oRVX23BBe3LI48Za7LejF7xPw&e=>

Reach out again if you need anything! ⚡ ⚡

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_Azure_azure-2Dfunctions-2Dhost_issues_6985-23issuecomment-2D796969693&d=DwMFaQ&c=eIGjsITfXP_y-DLLX0uEHXJvU8nOHrUK8IrwNKOtkVU&r=BbThQ-iApsLvCBycuzvhkO8FJNgUkO6tTOl2tJGWpHk&m=MqDA26oNjO8DUewvFmarqGTHJFy6DKFnZE15mMN7p2A&s=PvPo9TTaO-uiKI3UVDkh-xqJsJ9Yyhu2njCYYR8O-sw&e=>, or unsubscribe<https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_notifications_unsubscribe-2Dauth_ACIWIOHXHV2FIMIEBGOBSOTTDEATDANCNFSM4U43JUWA&d=DwMFaQ&c=eIGjsITfXP_y-DLLX0uEHXJvU8nOHrUK8IrwNKOtkVU&r=BbThQ-iApsLvCBycuzvhkO8FJNgUkO6tTOl2tJGWpHk&m=MqDA26oNjO8DUewvFmarqGTHJFy6DKFnZE15mMN7p2A&s=84oRTDxhyKAJwm8L65T6aBVJanzKrW_28NUA5Cncz0w&e=>.

________________________________

This message is for the designated recipient only and may contain privileged, proprietary, or otherwise confidential information. If you have received it in error, please notify the sender immediately and delete the original. Any other use of the e-mail by you is prohibited. Where allowed by local law, electronic communications with Accenture and its affiliates, including e-mail and instant messaging (including content), may be scanned by our systems for the purposes of information security and assessment of internal compliance with Accenture policy. Your privacy is important to us. Accenture uses your personal data only in compliance with data protection laws. For further information on how Accenture processes your personal data, please see our privacy statement at https://www.accenture.com/us-en/privacy-policy.
______________________________________________________________________________________

www.accenture.com
